samples/HttpServer_FirmwareUpload/app/application.cpp

```diff
@@ -3,10 +3,120 @@
 #include <Data/Stream/LimitedMemoryStream.h>
 #include <MultipartParser/HttpMultipartResource.h>
 #include <Data/Stream/RbootOutputStream.h>
+#include <sodium/crypto_sign.h>
+#include "FirmwareVerificationKey.h"
 
 HttpServer server;
 String lastModified;
 
+class SignedRbootOutputStream : public RbootOutputStream
```

Comment:
Take that class out of  application.cpp. You can move it to a separate file in this sample OR if you think it is good enough to be reused move it to `Components/libsodium` . Similar to [MultipartParser](https://github.com/SmingHub/Sming/tree/develop/Sming/Components/MultipartParser). I would recommend you to do the latter.

Reply:
Separate file it is. 
Moving it to libsodium isn't the right choice, because libsodium offers cryptographic functions, not firmware update functionality.

Reply:
> .. Learned a lot today!  ...
> .. but this should be done in a separate PR.

Wonderful! Thanks for the fantastic work so far.

Reply:
Thanks for your support as well!